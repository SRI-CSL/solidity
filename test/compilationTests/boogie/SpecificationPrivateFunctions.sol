pragma solidity ^0.4.23;

/**
 * @notice invariant x == y
 */
contract SpecificationPrivateFunctions {
    uint x;
    uint y;

    function add(uint amount) public {
        // Private functions do not have the invariant as pre/postcondition
        // so they should be inlined
        addToX(amount);
        addToY(amount);
    }

    function addWrong(uint amount) public {
        addToXWrong(amount); // ERROR
        addToY(amount);
    }

    function addToX(uint amount) private {
        x += amount;
    }

    function addToXWrong(uint amount) private {
        x -= amount;
    }

    function addToY(uint amount) private {
        y += amount;
    }
}