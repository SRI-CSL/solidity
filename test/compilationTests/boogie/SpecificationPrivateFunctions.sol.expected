Verifying SpecificationPrivateFunctions::add: OK
Verifying SpecificationPrivateFunctions::addWrong: ERROR
 - Source SpecificationPrivateFunctions.sol, line 17, col 5: Invariant 'x == y' might not hold at end of function.
